src/core/coap/coap.cpp

```diff
@@ -601,6 +601,7 @@ void CoapBase::ProcessReceivedRequest(Message &aMessage, const Ip6::MessageInfo
     switch (mResponsesQueue.GetMatchedResponseCopy(aMessage, aMessageInfo, &cachedResponse))
     {
     case OT_ERROR_NONE:
+        cachedResponse->Finish();
```

Comment:
Great catch! Should this be inside `GetMatchedResponseCopy()`?

Reply:
The enqueued responses are not `Finish`ed. 
If `GetMatchedResponseCopy` returns messages already `Finish`ed, wouldn't that be asymmetric?

Reply:
Should the enqueued responses be "finished" in `EnqueueResponse()`?

Reply:
Yes, I think so. However, this PR was meant to be a quick fix for this bug.
@bukepo is working on a more comprehensive refactoring of coap message sending code.